File: frontend/src/api/monitor.ts
import request from '@/utils/request'

// 告警策略 Types
export interface AlertPolicy {
  id: number
  name: string
  status: string
  enabled: boolean
  resource_type: string
  metric: string
  threshold: number
  duration: number
  level: string
  owner: string
  domain_id: number
  project_id: number
  description: string
  notify_channel: string
  created_at: string
  updated_at: string
}

export interface AlertPolicyRequest {
  name: string
  resource_type: string
  metric: string
  threshold: number
  duration?: number
  level?: string
  enabled?: boolean
  owner?: string
  domain_id?: number
  project_id?: number
  description?: string
  notify_channel?: string
}

export interface AlertPolicyFilter {
  resource_type?: string
  owner?: string
  level?: string
  enabled?: string
}

// 告警历史 Types
export interface AlertHistory {
  id: number
  policy_id: number
  policy_name: string
  resource_id: string
  resource_name: string
  resource_type: string
  level: string
  status: string
  metric_value: number
  message: string
  triggered_at: string
  resolved_at?: string
  domain_id: number
  project_id: number
}

export interface AlertHistoryFilter {
  status?: string
  level?: string
  resource_type?: string
  start_time?: string
  end_time?: string
  page?: number
  page_size?: number
}

// 监控资源 Types
export interface MonitorResource {
  id: number
  resource_id: string
  resource_name: string
  resource_type: string
  monitor_status: string
  account_id: number
  platform: string
  region: string
  project_id: number
  last_sync_at: string
  metrics: string
}

export interface ResourceMetrics {
  resource_id: string
  timestamp: number
  cpu_usage: { value: number; unit: string; timestamp: number }
  memory_usage: { value: number; unit: string; timestamp: number }
  disk_usage: { value: number; unit: string; timestamp: number }
  network_in: { value: number; unit: string; timestamp: number }
  network_out: { value: number; unit: string; timestamp: number }
}

// 告警策略 API
export function listAlertPolicies(filter?: AlertPolicyFilter) {
  return request.get<AlertPolicy[]>('/monitor/alert-policies', { params: filter })
}

export function createAlertPolicy(data: AlertPolicyRequest) {
  return request.post<AlertPolicy>('/monitor/alert-policies', data)
}

export function getAlertPolicy(id: number) {
  return request.get<AlertPolicy>(`/monitor/alert-policies/${id}`)
}

export function updateAlertPolicy(id: number, data: AlertPolicyRequest) {
  return request.put<AlertPolicy>(`/monitor/alert-policies/${id}`, data)
}

export function deleteAlertPolicy(id: number) {
  return request.delete(`/monitor/alert-policies/${id}`)
}

export function toggleAlertPolicy(id: number) {
  return request.post<AlertPolicy>(`/monitor/alert-policies/${id}/toggle`)
}

// 告警历史 API
export function listAlertHistory(filter?: AlertHistoryFilter) {
  return request.get<{ data: AlertHistory[]; total: number; page: number; page_size: number }>('/monitor/alert-history', { params: filter })
}

export function resolveAlert(id: number) {
  return request.post<AlertHistory>(`/monitor/alert-history/${id}/resolve`)
}

export function ignoreAlert(id: number) {
  return request.post<AlertHistory>(`/monitor/alert-history/${id}/ignore`)
}

// 监控资源 API
export function listMonitorResources(accountId?: number, resourceType?: string, monitorStatus?: string) {
  return request.get<MonitorResource[]>('/monitor/resources', {
    params: {
      account_id: accountId,
      resource_type: resourceType,
      monitor_status: monitorStatus
    }
  })
}

export function syncMonitorResources(accountId: number) {
  return request.post<{ message: string; count: number }>('/monitor/resources/sync', null, {
    params: { account_id: accountId }
  })
}

export function getResourceMetrics(accountId: number, resourceId: string) {
  return request.get<ResourceMetrics>(`/monitor/resources/${resourceId}/metrics`, {
    params: { account_id: accountId }
  })
}

File: backend/internal/model/monitor.go
package model

import (
	"time"

	"gorm.io/gorm"
)

// AlertPolicy 告警策略
type AlertPolicy struct {
	gorm.Model
	Name          string         `gorm:"size:100;not null" json:"name"`
	Status        string         `gorm:"size:20;default:'正常'" json:"status"`
	Enabled       bool           `gorm:"default:true" json:"enabled"`
	ResourceType  string         `gorm:"size:50" json:"resource_type"` // 虚拟机/数据库/网络等
	Metric        string         `gorm:"size:50" json:"metric"`        // cpu_usage/memory_usage等
	Threshold     float64        `gorm:"type:decimal(10,2)" json:"threshold"`
	Duration      int            `gorm:"default:5" json:"duration"` // 持续时间(分钟)
	Level         string         `gorm:"size:20" json:"level"`      // 信息/警告/严重
	Owner         string         `gorm:"size:50" json:"owner"`      // 系统/自定义
	DomainID      uint           `json:"domain_id"`
	ProjectID     uint           `json:"project_id"`
	ResourceTags  string         `gorm:"size:500" json:"resource_tags"` // JSON格式的标签匹配规则
	Description   string         `gorm:"size:500" json:"description"`
	NotifyChannel string         `gorm:"size:100" json:"notify_channel"` // 通知渠道
	CreatedAt     time.Time      `json:"created_at"`
	UpdatedAt     time.Time      `json:"updated_at"`
}

// AlertHistory 告警历史
type AlertHistory struct {
	gorm.Model
	PolicyID     uint      `json:"policy_id"`
	PolicyName   string    `gorm:"size:100" json:"policy_name"`
	ResourceID   string    `gorm:"size:100" json:"resource_id"`
	ResourceName string    `gorm:"size:100" json:"resource_name"`
	ResourceType string    `gorm:"size:50" json:"resource_type"`
	Level        string    `gorm:"size:20" json:"level"`
	Status       string    `gorm:"size:20" json:"status"` // pending/resolved/ignored
	MetricValue  float64   `gorm:"type:decimal(10,2)" json:"metric_value"`
	Message      string    `gorm:"size:500" json:"message"`
	TriggeredAt  time.Time `json:"triggered_at"`
	ResolvedAt   *time.Time `json:"resolved_at"`
	DomainID     uint      `json:"domain_id"`
	ProjectID    uint      `json:"project_id"`
}

// MonitorResource 监控资源
type MonitorResource struct {
	gorm.Model
	ResourceID     string    `gorm:"size:100;uniqueIndex" json:"resource_id"`
	ResourceName   string    `gorm:"size:100" json:"resource_name"`
	ResourceType   string    `gorm:"size:50" json:"resource_type"` // vm/rds/redis等
	MonitorStatus  string    `gorm:"size:20" json:"monitor_status"` // 正常/告警/异常
	AccountID      uint      `json:"account_id"`
	Platform       string    `gorm:"size:50" json:"platform"`
	Region         string    `gorm:"size:50" json:"region"`
	ProjectID      uint      `json:"project_id"`
	LastSyncAt     time.Time `json:"last_sync_at"`
	Metrics        string    `gorm:"type:text" json:"metrics"` // JSON格式的最新指标数据
}